builtin/providers/aws/resource_aws_sns_platform_application.go

```diff
@@ -0,0 +1,192 @@
+
+	return resourceAwsSnsPlatformApplicationAPNSRead(d, meta)
+}
+
+func resourceAwsSnsPlatformApplicationAPNSRead(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	_, err := snsconn.GetPlatformApplicationAttributes(&sns.GetPlatformApplicationAttributesInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+	})
+
+	if err != nil {
+		return fmt.Errorf("Error reading plaform application %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsSnsPlatformApplicationAPNSUpdate(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	resp, err := snsconn.SetPlatformApplicationAttributes(&sns.SetPlatformApplicationAttributesInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+		Attributes: aws.StringMap(map[string]string{
+			"PlatformCredential": d.Get("platform_credential").(string),
+			"PlatformPrincipal":  d.Get("platform_principal").(string),
+		}),
+	})
+
+	if err != nil {
+		return fmt.Errorf("Error reading platform application %s", err)
+	}
+
+	log.Printf("[DEBUG] Received GCM platform application: %s", resp)
+
+	return nil
+}
+
+func resourceAwsSnsPlatformApplicationDelete(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	_, err := snsconn.DeletePlatformApplication(&sns.DeletePlatformApplicationInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+	})
+
+	if err != nil {
+		return fmt.Errorf("Error deleting plaform application %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsSnsPlatformApplicationGCM() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSnsPlatformApplicationGCMCreate,
+		Read:   resourceAwsSnsPlatformApplicationGCMRead,
+		Update: resourceAwsSnsPlatformApplicationGCMUpdate,
+		Delete: resourceAwsSnsPlatformApplicationDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"platform_credential": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceAwsSnsPlatformApplicationGCMCreate(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	resp, err := snsconn.CreatePlatformApplication(&sns.CreatePlatformApplicationInput{
+		Name:     aws.String(d.Get("name").(string)),
+		Platform: aws.String("GCM"),
+		Attributes: aws.StringMap(map[string]string{
+			"PlatformCredential": d.Get("platform_credential").(string),
+		}),
+	})
+	if err != nil {
+		return fmt.Errorf("Error reading platform application %s", err)
+	}
+
+	d.SetId(*resp.PlatformApplicationArn)
+
+	return resourceAwsSnsPlatformApplicationGCMRead(d, meta)
+}
+
+func resourceAwsSnsPlatformApplicationGCMRead(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	_, err := snsconn.GetPlatformApplicationAttributes(&sns.GetPlatformApplicationAttributesInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+	})
+
+	if err != nil {
```

Comment:
There are cases where terraform creates infra, someone deletes it from the UI and then this will fail. In this instance, you may want to check that if the err is the equivalent of a 404, then to log it, d.SetId("") and then return nil as per this style example https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_ecr_repository.go#L68